pkg/migrator/cleanup.go

```diff
@@ -0,0 +1,246 @@
+// are updated with information about datastore CatalogSources
+// instead.
+func (u *upgrader) UpdateSubscriptions() ([]cscInfo, error) {
+	var installedCscs []cscInfo
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf(builders.OwnerNameLabel))
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing subscriptions that have the label "csc-owner-name"
+	err := u.client.List(context.TODO(), options, subscriptions)
+	if err != nil {
+		return nil, err
+	}
+	for _, instance := range subscriptions.Items {
+		newCscInfo := cscInfo{name: instance.GetLabels()[builders.OwnerNameLabel], namespace: instance.GetLabels()[builders.OwnerNamespaceLabel]}
+		installedCscs = append(installedCscs, newCscInfo)
+		// try to infer the datastore catalogsource from the subscription
+		datastoreCs, err := findCatalogSource(&instance, u.client)
+		if k8s_errors.IsNotFound(err) {
+			// guess the catalogsource from the opsrc that has the package
+			datastoreCs, err = findCsFromOpsrc(&instance, u.client)
+			if err != nil {
+				u.logger.Errorf("[migration] Could not infer datastore CatalogSource for Subscription %s.", instance.GetName())
+				continue
+			}
+		}
+		// point the subscription to the datastore CatalogSource
+		updatedSub := setSubsCatalogSource(&instance, datastoreCs.GetName(), datastoreCs.GetNamespace())
+		err = u.client.Update(context.TODO(), updatedSub)
+		if err != nil {
+			u.logger.Errorf("[migration] Error updating subscription %s. Error: %s", instance.GetName(), err.Error())
+		} else {
+			u.logger.Println("[migration] Succesfully updated Subscription %s", instance.GetName())
+		}
+	}
+	return installedCscs, nil
+}
+
+// deleteInstalledCscs deletes the CSCs installed during opeartor installation.
+// The child resources of the CSCs are delete by the finalizer.
+func deleteInstalledCscs(cscs []cscInfo, kubeClient client.Client, logger *logrus.Entry) error {
+	for _, cscInfo := range cscs {
+		csc := helperNewCatalogSourceConfig(cscInfo.namespace, cscInfo.name)
+		err := kubeClient.Delete(context.TODO(), csc)
+		if err != nil {
+			logger.Errorf("[migration] Failed to delete installed CSC %s with error: ", cscInfo.name, err.Error())
+		} else {
+			logger.Printf("[migration] Succesfully deleted stale CSC %s,", cscInfo.name)
+		}
+	}
+	return nil
+}
+
+// deleteDatastoreCscs deletes the datastore CSCs created by OperatorSources.
+// The child resources of the CSCs are delete by the finalizer.
+func deleteDatastoreCscs(kubeClient client.Client, logger *logrus.Entry) error {
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("opsrc-datastore: \"true\""))
+
+	cscs := &v2.CatalogSourceConfigList{}
+	// Get the list of existing cscs that have the lable "opsrc-datastore: "true" "
+	err := kubeClient.List(context.TODO(), options, cscs)
+	if err != nil {
+		return err
+	}
+	for _, csc := range cscs.Items {
+		err = kubeClient.Delete(context.TODO(), &csc)
+		if err != nil {
```

Comment:
Why don't we ever try to return this error? I understand the desire to try on every item before we give up, but we could aggregate the error and return a list of errors at the end, or even aggregate the errors into a single error when returning.